pkg/k8shandler/cluster.go

```diff
@@ -47,183 +46,150 @@ func (elasticsearchRequest *ElasticsearchRequest) CreateOrUpdateElasticsearchClu
-			logrus.Warnf("Unable to clear transient shard allocation for cluster %q in namespace %q: %s", elasticsearchRequest.cluster.Namespace, elasticsearchRequest.cluster.Namespace, err.Error())
-		}
-	}
+	er.ensureNoTransitiveShardAllocations()
 
-	progressUnschedulableNodes(elasticsearchRequest.cluster)
+	if err := er.progressUnschedulableNodes(); err != nil {
+		logrus.Warnf("unable to progress unschedulable nodes for %v in namespace %v", er.cluster.Name, er.cluster.Namespace)
+		return er.UpdateClusterStatus()
+	}
 
-	certRestartNodes := getScheduledCertRedeployNodes(elasticsearchRequest.cluster)
-	stillRecovering := containsClusterCondition(api.Recovering, v1.ConditionTrue, &elasticsearchRequest.cluster.Status)
+	certRestartNodes := er.getScheduledCertRedeployNodes()
+	stillRecovering := containsClusterCondition(api.Recovering, v1.ConditionTrue, &er.cluster.Status)
 	if len(certRestartNodes) > 0 || stillRecovering {
-		err = elasticsearchRequest.PerformFullClusterCertRestart(certRestartNodes)
-		if err != nil {
-			return elasticsearchRequest.UpdateClusterStatus()
+		if err := er.PerformFullClusterCertRestart(certRestartNodes); err != nil {
+			logrus.Warnf("unable to complete full cluster restart for %v in namespace %v", er.cluster.Name, er.cluster.Namespace)
+			return er.UpdateClusterStatus()
 		}
+
+		er.UpdateClusterStatus()
 	}
 
 	// if there is a node currently being upgraded, work on that first
-	inProgressNode := getNodeUpgradeInProgress(elasticsearchRequest.cluster)
-	scheduledNodes := getScheduledUpgradeNodes(elasticsearchRequest.cluster)
+	inProgressNode := er.getNodeUpgradeInProgress()
+	scheduledNodes := er.getScheduledUpgradeNodes()
 
-	// Check if we have a node that was in the progress of an update
-	// if so, continue updating it
+	// Check if we have a node that was in the progress -- if so, continue updating it
 	if inProgressNode != nil {
-		// currently no way to distinguish between the two -- restart node and update node
-		//  however we don't currently have anything that will trigger just a rolling restart of nodes
-		//if _, ok := containsNodeTypeInterface(inProgressNode, scheduledUpgradeNodes); ok {
-		logrus.Debugf("Continuing update for %v", inProgressNode.name())
-		if err := elasticsearchRequest.PerformNodeUpdate(inProgressNode); err != nil {
-			logrus.Warnf("unable to update node. E: %s", err.Error())
-		}
-		/*} else {
+		// Check to see if the inProgressNode was being updated or restarted
+		if _, ok := containsNodeTypeInterface(inProgressNode, scheduledNodes); ok {
+			logrus.Debugf("Continuing update for %v", inProgressNode.name())
+			if err := er.PerformNodeUpdate(inProgressNode); err != nil {
+				logrus.Warnf("unable to update node. E: %s", err.Error())
+				return er.UpdateClusterStatus()
+			}
+
+			// update scheduled nodes since we were able to complete upgrade for inProgressNode
+			scheduledNodes = er.getScheduledUpgradeNodes()
```

Comment:
Was this previously a bug? Or is this a consequence of enabling the commented conditional now?

Reply:
since we don't have nested if blocks when we complete here it is possible for us to continue to update the next node in the original list of nodes that are scheduled to upgrade. to enter this current block of code `inProgressNode` needs to be within the list of nodes in `scheduledNodes`. so if we've finished updating `inProgressNode` then we want to update the list of nodes to be scheduled to no longer contain `inProgressNode` so it won't attempt to be updated again below.

Reply:
IC, this is of sort `inProgressNode` is done, so drop it from the list by updating the list `scheduledNodes`. I don't quite get which "nested if blocks" we don't have anymore. In total it makes sense to me now.